### PR TITLE
Fix #1409 (assert when trying to get one value as multiple types)

### LINKED ISSUE #1409: ondemand::parser::get command reordering leads to segfault
This simple example causes an segfault.
The first loop-run works, the second causes a segfault
switch the two get statements leads to an segfault in the first run.
Can somebody reproduce this behaviour?

Tested on:
Linux ubuntu 20.04 gcc9 and gcc10
Windows 10 64 bit gcc9
CPU: amd and intel

```
#include <singleheader/simdjson.h>

int main()
{
    using namespace simdjson;
    auto const abstract_json= R"(
                              { "str" : 123 }
                              )"_padded;

    for ( int i= 0; i < 2; ++i )
    {
        ondemand::parser parser;
        auto doc = parser.iterate( abstract_json );
        ondemand::value val= doc["str"];
        if ( i == 0 )
        {
            auto a= val.get<ondemand::array>();
            auto s= val.get<std::string_view>();
        }
        else
        {
            auto s= val.get<std::string_view>();
            auto a= val.get<ondemand::array>(); //<-- segfault
        }
#if 0
        auto d= val.get<double>();
        auto u= val.get<uint64_t>();
        auto i= val.get<int64_t>();
        auto b= val.get<bool>();
        auto o= val.get<ondemand::object>();
#endif
    }
    return 0;
}
```

### PR DESCRIPTION
**Pull request by jkeiser**
This fixes problems when you try to run get_bool(), get_array(), etc. against the same value: arrays, objects and null were particularly doing the wrong thing. In so doing, it also starts parsing arrays and objects the same way as scalars: even if the type is wrong, the iterator is automatically advanced to the next token.

It also makes failed number parsing return INCORRECT_TYPE, so that you can switch on that to try again.

Fixes #1409.

### CODE REVIEW

**Comment by jkeiser** [comment]
This gives me a 3% degradation in performance on find_tweets--possibly others. I can't figure out where the issue would be, but simpler changes than this give me the same degradation.

**Comment by jkeiser** [comment]
I'd like to spend some time figuring out where the 3% went; the change in the jkeiser/safe-iterators branch causes the same problem, so they may have the same cause.

Regardless, this is necessary for the bugfix, and I am hoping they *do* have the same cause, so I can make safe iterators without any further cost :)

**Comment by lemire** [comment]
The requests for reviews have been noted btw.

File: include/simdjson/generic/numberparsing.h [hunk 1]
```diff
@@ -608,12 +608,15 @@ simdjson_unused simdjson_really_inline simdjson_result<uint64_t> parse_unsigned(
 
   // If there were no digits, or if the integer starts with 0 and has more than one digit, it's an error.
   int digit_count = int(p - start_digits);
-  if (digit_count == 0 || ('0' == *start_digits && digit_count > 1)) { return NUMBER_ERROR; }
-  if (!jsoncharutils::is_structural_or_whitespace(*p)) { return NUMBER_ERROR; }
+  if (digit_count == 0) { return INCORRECT_TYPE; }
+  if ('0' == *start_digits && digit_count > 1) { return NUMBER_ERROR; }
+  if (!jsoncharutils::is_structural_or_whitespace(*p)) {
```
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:613]: Do we want simdjson_unlikely around this check?  What is inside the 'if' clause is not going to be cheap with this change. (May not help.)

File: include/simdjson/generic/numberparsing.h [hunk 2]
```diff
@@ -608,12 +608,15 @@ simdjson_unused simdjson_really_inline simdjson_result<uint64_t> parse_unsigned(
...
+    return (*p == '.' || *p == 'e' || *p == 'E') ? INCORRECT_TYPE : NUMBER_ERROR;
+  }
 
   // The longest positive 64-bit number is 20 digits.
   // We do it this way so we don't trigger this branch unless we must.
-  if (digit_count > 20) { return NUMBER_ERROR; }
+  if (digit_count > 20) { return INCORRECT_TYPE; }
```
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:619]: Can this be merged with 
```
if (digit_count == 0) { return INCORRECT_TYPE; }
```
?
The following should be faster...

```
if ((digit_count <= 0) ||(digit_count > 20)) { return INCORRECT_TYPE; }
```
Because... you only have to check that `unsigned(digit_count-1)>=20`. So just one branch,

File: include/simdjson/generic/numberparsing.h (hunk 1 above)
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:613]: is_not_structural_or_whitespace might be faster here (saves one instruction).

Comment (lemire) [reply include/simdjson/generic/numberparsing.h:613]: It might pay here to have a dedicated table...

```
static error_code * can_end_int =....
if(can_end_int[p]!=SUCCESS) { return can_end_int[p];}
```

File: include/simdjson/generic/numberparsing.h [hunk 3]
```diff
@@ -651,18 +654,21 @@ simdjson_unused simdjson_really_inline simdjson_result<int64_t> parse_integer(co
 
   // If there were no digits, or if the integer starts with 0 and has more than one digit, it's an error.
   int digit_count = int(p - start_digits);
-  if (digit_count == 0 || ('0' == *start_digits && digit_count > 1)) { return NUMBER_ERROR; }
-  if (!jsoncharutils::is_structural_or_whitespace(*p)) { return NUMBER_ERROR; }
+  if (digit_count == 0) { return INCORRECT_TYPE; }
+  if ('0' == *start_digits && digit_count > 1) { return NUMBER_ERROR; }
+  if (!jsoncharutils::is_structural_or_whitespace(*p)) {
+    return (*p == '.' || *p == 'e' || *p == 'E') ? INCORRECT_TYPE : NUMBER_ERROR;
+  }
```
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:661]: Same story... it might pay to gave a dedicated table that returns the error_code: we need or success. We do a lookup anyhow, we might as well get the error code for free? Fewer branches and instructions at the cost of a new small table.

File: include/simdjson/generic/numberparsing.h [hunk 4]
```diff
@@ -651,18 +654,21 @@ simdjson_unused simdjson_really_inline simdjson_result<int64_t> parse_integer(co
...
 
   // The longest negative 64-bit number is 19 digits.
   // The longest positive 64-bit number is 20 digits.
   // We do it this way so we don't trigger this branch unless we must.
   int longest_digit_count = negative ? 19 : 20;
-  if (digit_count > longest_digit_count) { return NUMBER_ERROR; }
+  if (digit_count > longest_digit_count) { return INCORRECT_TYPE; }
```
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:667]: same story here, I think we should merge he two digit_count check and make it so the compiler only needs one branch.

File: include/simdjson/generic/numberparsing.h [hunk 5]
```diff
@@ -608,12 +608,15 @@ simdjson_unused simdjson_really_inline simdjson_result<uint64_t> parse_unsigned(
 
   // If there were no digits, or if the integer starts with 0 and has more than one digit, it's an error.
   int digit_count = int(p - start_digits);
-  if (digit_count == 0 || ('0' == *start_digits && digit_count > 1)) { return NUMBER_ERROR; }
-  if (!jsoncharutils::is_structural_or_whitespace(*p)) { return NUMBER_ERROR; }
+  if (digit_count == 0) { return INCORRECT_TYPE; }
```
Comment (lemire) [inline include/simdjson/generic/numberparsing.h:611]: it might pay to use an unsigned type for digit_count to ease optimizations

**Comment by lemire** [comment]
@jkeiser I will see about producing a PR against this PR that might have different performance characteristics (maybe better).

Note: this is good code as usual, but 3% is bad. Let me see...

**Review by lemire — CHANGES_REQUESTED** [review]
Please review PR https://github.com/simdjson/simdjson/pull/1421 before merging this.

I am marking this as "request changes" not to block it, but to make sure my PR is considered.

**Comment by lemire** [comment]
My quick tests on my PR suggest that the performance damage can be fully recovered.

**Review by lemire — APPROVED** [review]
We may want to re-assess the performance before merging.

My guess is that this PR is now performance neutral (roughly).

**Comment by jkeiser** [comment]
@lemire thanks again! Performance is sitting at 5.85GB/s versus 5.95GB/s for master. find_tweet does very little number parsing, so the low impat makes some sense.

I still need to take a look at *why*, but I'm also considering reenabling SIMDJSON_ONDEMAND_SAFETY_RAILS on this sucker. Let me give that a try as well (renamed to SIMDJSON_API_USAGE_CHECKS). It seems prudent to turn it on by default, and include a SIMDJSON_PRODUCTION flag that turns it off as well ("we recommend you use this after you have finished developing and testing your software").

We still can't let it be much slower than production mode, and I don't think it will be--but we may need a 5% hit to ensure that people can write their software and be confident it works no matter what.

**Comment by jkeiser** [comment]
Oh ... this isn't one of the actual out of order iteration checks. All right, I'll get back to tracking this down.

**Comment by lemire** [comment]
Our old perf test fails (DOM based). Locally, I find that this PR adds about 1% to the DOM parse instruction count, while it is neutral for GNU GCC. So it is fine as far as that goes. I am not sure why we have a CI failure.

**Comment by lemire** [comment]
Let me correct that: up to a 1% increase in instruction count under clang. Not 1% across the board. We get a 1% increase with files like twitter.json but much less of an impact with canada.json. Under GCC, this PR is generally beneficial, reducing instruction count by up to 0.4% (mesh.json). So we are up (in instruction count) on one side and down on the other. Running times go both up and down (+/- 5%).

To recap, as far as DOM parsing goes, this PR is not affecting performance as far as I can see.

**Comment by jkeiser** [comment]
Ok! That is a good reason to keep it. I need to adjust my performance spot checking, too, and compare both compilers.

I have an experiment almost ready that does this with a lighter touch, just to establish which part of the change is affecting things. But given this I probably won't go further than an experiment.

**Comment by jkeiser** [comment]
OK! ... I can't say I understand why DOM is taking a hit ... write_float() is part of DOM too, yes? Maybe the change from size_t to int is hurting it because it's inserting a conversion somewhere?

Anyway, my experiments with a lighter touch have the same degradation, so I think it's inevitable. I even added more assumptions in case clang just couldn't tell where the cursor was. (I like the word cursor enough I think I'm going to rename value_iterator and json_iterator to value_cursor. It conveys the flavor of what's going on much better.)
